staging/src/k8s.io/component-base/featuregate/feature_gate.go

```diff
@@ -475,6 +501,115 @@ func (f *featureGate) AddVersioned(features map[Feature]VersionedSpecs) error {
+	known := f.GetAllVersioned()
+
+	// Merge existing dependencies in.
+	existing := *f.dependencies.Load()
+	for k, v := range existing {
+		dependencies[k] = append(dependencies[k], v...)
+	}
+
+	// Sort and compact all dependency lists.
+	for k, v := range dependencies {
+		slices.Sort(v)
+		dependencies[k] = slices.Compact(v)
+	}
+
+	// Features cannot depend on features with a lower prerelease level.
+	// This function converts prereleases to a numerical value for comparison.
+	stabilityOrder := func(p prerelease) int {
+		switch p {
+		case Deprecated:
+			return 0 // Non-deprecated features cannot depend on deprecated features.
+		case PreAlpha:
+			return 1
+		case Alpha:
+			return 2
+		case Beta:
+			return 3
+		case GA:
+			return 4
+		default:
+			return -1 // Unknown prerelease
+		}
+	}
+
+	// Validate dependencies for each emulated version:
+	// 1. Features & dependencies must be known
+	// 2. Features cannot depend on features with a lower prerelease level
+	// 3. Enabled features cannot depend on disabled features
+	// 4. Locked-to-default featurs cannot depend on unlocked features
+	for feature, deps := range dependencies {
+		versionedFeature, ok := known[feature]
+		if !ok {
+			return fmt.Errorf("cannot add dependency for unknown feature %s", feature)
+		}
+
+		for _, dep := range deps {
+			versionedDep, ok := known[dep]
+			if !ok {
+				return fmt.Errorf("cannot add dependency from %s to unknown feature %s", feature, dep)
+			}
+
+			// Check versions starting with the most recent for more intuitive error messages.
+			for _, spec := range slices.Backward(versionedFeature) {
+				// depSpec is the effective FeatureSpec for the dependency at the version declared by the dependent FeatureSpec.
+				depSpec := featureSpecAtEmulationVersion(versionedDep, spec.Version)
+
+				if stabilityOrder(spec.PreRelease) > stabilityOrder(depSpec.PreRelease) {
+					return fmt.Errorf("%s feature %s cannot depend on %s feature %s at version %s", spec.PreRelease, feature, depSpec.PreRelease, dep, spec.Version.String())
+				}
+				if spec.Default && !depSpec.Default {
+					return fmt.Errorf("default-enabled feature %s cannot depend on default-disabled feature %s at version %s", feature, dep, spec.Version.String())
+				}
+				if spec.LockToDefault && !depSpec.LockToDefault {
+					return fmt.Errorf("locked-to-default feature %s cannot depend on unlocked feature %s at version %s", feature, dep, spec.Version.String())
+				}
+			}
+		}
+	}
+
+	// Check for cycles
+	visited := map[Feature]bool{}
+	finished := map[Feature]bool{}
+	var detectCycles func(Feature) error
+	detectCycles = func(feature Feature) error {
+		if finished[feature] {
+			return nil
+		}
+		if visited[feature] {
+			return fmt.Errorf("cycle detected with feature %s", feature)
+		}
+		visited[feature] = true
+		for _, dep := range dependencies[feature] {
+			if err := detectCycles(dep); err != nil {
+				return err
+			}
+		}
+		finished[feature] = true
+		return nil
+	}
```

Comment:
How do we expect to handle transitive dependencies? 

If we have three features A, B, C, with dependencies as
```
A : [B,C]
B : [C]
C : []
```
Then if we run through A's dependencies, we are going to visit C while detecting cycles in B, and then visit C again when we continue the range over A's dependencies

At which case i think this would report a cycle, but I don't think that is correct?

AFAICT there isn't a test case demonstrating this kind of relationship at the moment

Reply:
I added this case to the test to confirm it works.

Here's how it goes, if this makes sense:
```
A -> visited(A); check B
B -> visited(B); check C
C -> visited(C); done(C)
B -> done(B)
A -> check (C), already done; done(A)
```
Basically when A checks C, C has been marked as done, so it doesn't flag a cycle. Being marked done means there is no path back to the node from it's own dependencies. C doesn't have deps, so it's immediately marked done.

See https://en.wikipedia.org/wiki/Cycle_(graph_theory)#Cycle_detection

Reply:
Ahh, I missed the purpose of the `done` part, thanks for the explanation